File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    map<int,int> hashMap; 
    vector<int> container; 
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(hashMap.count(val)) return false; 

        container.push_back(val); 
        hashMap[val] = container.size()-1; 

        return true; 
    }
    
    bool remove(int val) {
        if(!hashMap.count(val)) return false;

        int index = hashMap[val]; 
        int last = container.back(); 

        container[index] = last; 
        container[container.size()-1] = val; 
        container.pop_back(); 

        hashMap.erase(val);
        hashMap[last] = index; 

        return true;         
    }
    
    int getRandom() {
        return container[rand() % container.size()]; 
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
class RandomizedSet {

    List<Integer> arrayList = new ArrayList<>(); 
    Map<Integer,Integer> hashMap = new HashMap<>(); 
    int index = 0; 

    public RandomizedSet() {
        
    }
    
    public boolean insert(int val) {
        if(hashMap.containsKey(val)) return false; 

        arrayList.add(val); 
        hashMap.put(val, index++); 

        return true; 
    }
    
    public boolean remove(int val) {
        if(!hashMap.containsKey(val)) return false; 

        int targetVal = arrayList.get(arrayList.size()-1); 
        int valIndex = hashMap.get(val); 

        arrayList.set(valIndex, targetVal); 
        arrayList.set(arrayList.size()-1, val); 

        arrayList.removeLast(); 
        hashMap.put(targetVal, valIndex); 
        hashMap.remove(val); 
        index--; 

        return true; 
    }
    
    public int getRandom() {
        int randomIndex = (int)(Math.random() * arrayList.size()); 
        return arrayList.get(randomIndex);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */